### PR TITLE
Only consider netlink "device" interfaces in address autodetection on Linux

### LINKED ISSUE #25139: Ignore bridge interfaces in address autodetection
Based on the discussion in #25081, it sounds like there is a consensus that bridge interfaces should not prevent swarm IP address autodetection from succeeding.

If the only IP address on the system is associated with a bridge interface, we should use that one. But if there is also an IP address on a "normal" interface, the bridge should be ignored.

This will probably require the use of platform-specific code (netlink).

### PR DESCRIPTION
**Pull request by aboch**
- This automatically rules out bridges and other non system
  created interfaces
- Retained the current logic on non Linux systems

Fixes #25139

Tested on ubuntu 3.13 and 4.2, 4.4

Signed-off-by: Alessandro Boch aboch@docker.com

### CODE REVIEW

**Comment by aboch** [comment]
ping @aaronlehmann

**Comment by thaJeztah** [comment]
@aboch @aaronlehmann which milestone should this be added to?

**Comment by aaronlehmann** [comment]
@thaJeztah: It feels like an enhancement, so I'd say 1.13 by default. But if we're seeing a lot of autodetection problems (and confused users) in the wild due to bridge interfaces, I wouldn't object to having it in 1.12.1.

cc @aluzzardi

File: daemon/cluster/listen_addr_linux.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+	"github.com/vishvananda/netlink"
+)
+
+func (c *Cluster) resolveSystemAddr() (net.IP, error) {
+	// Use the system's only device IP address, or fail if there are
+	// multiple addresses to choose from.
+	interfaces, err := netlink.LinkList()
+	if err != nil {
+		return nil, err
```
Comment (aaronlehmann) [inline daemon/cluster/listen_addr_linux.go]: Would it make sense to fall back to the generic detection function if `LinkList` fails? It looks like Go's `net.Interfaces` function uses netlink on Linux though, so perhaps there's no situation where this will fail and `net.Interfaces` will succeed.

File: daemon/cluster/listen_addr_linux.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		} else if interfaceAddr6 != nil {
+			if systemAddr != nil {
+				return nil, errMultipleIPs
+			}
+			systemAddr = interfaceAddr6
+		}
+	}
+
+	if systemAddr == nil {
+		return nil, errNoIP
```
Comment (aaronlehmann) [inline daemon/cluster/listen_addr_linux.go]: Instead of erroring out if there's no IP on a "device" interface, how about repeating the loop, and this time considering non-"device" interfaces as well. That way, a system that has a single bridge interface will support autodetection.

We could factor out the `for _, intf := range interfaces {` loop into a function that takes a boolean argument controlling whether to consider only device interfaces. The first attempt would only consider devices, and if no usable IP was found in that pass, the function would be called again with the "consider all interfaces" flag set to true.

**Comment by aaronlehmann** [comment]
Needs a rebase.

File: daemon/cluster/listen_addr_linux.go (hunk 1 above)
Comment (aboch) [inline daemon/cluster/listen_addr_linux.go]: Yes, net.Interfaces() on Linux also uses netlink

File: daemon/cluster/listen_addr_linux.go (hunk 2 above)
Comment (aboch) [inline daemon/cluster/listen_addr_linux.go]: We could do it, but wouldn't that be even more confusing to explain if a user actually ends is such a situation ?

In my view the autodetection is for covering the regular 99% of cases, where a physical device provides the connectivity to outside the host, and that was the reason behind this change.
Therefore it seems reasonable to me to have the autodetection fail if it cannot find an "eth" like device. Also I feel it is something people would find natural, and would be comfortable with it, in case of failure.

If we add self-healing logic, I am afraid it gets complicated, both to explain and even more to maintain. For example, I am not convinced we should choose a bridge interface as second external connectivity choice, rather than an ipvaln interface, for example. I feel it could easily take us into a rat hole of a priority list for backup interface types.

In summary, I'd prefer a simple predictable logic which covers the majority of the users' regular cases.

But of course if there is a strong argument in favor of a more intelligent autodetection mechanism and it turns out there are many use cases to support it, we should do it.

Comment (aaronlehmann) [inline daemon/cluster/listen_addr_linux.go]: I'm not arguing for an actual priority list. I just think that if a user only has a `br0` interface, and we don't choose it, that's strange behavior. If a user gets into this situation, I think it would be confusing to explain why we don't pick `br0` under the current code in this PR, but it wouldn't be confusing at all if we do pick it, because it would just work.

If there are multiple non-device interfaces with IP addresses, autodetection should fail, as before. This avoids any of the priority list rathole.

File: daemon/cluster/listen_addr_linux.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+			ipAddr := addr.IPNet.IP
+
+			// Skip loopback and link-local addresses
+			if !ipAddr.IsGlobalUnicast() {
+				continue
+			}
+
+			if ipAddr.To4() != nil {
+				if interfaceAddr4 != nil {
+					return nil, fmt.Errorf("could not choose an IP address to advertise since this system has multiple addresses on interface %s (%s and %s)",
```
Comment (aaronlehmann) [inline daemon/cluster/listen_addr_linux.go]: These errors are being created in a lot of places now that there are two implementations of `resolveSystemAddr`. How about introducing a function like:

```go
func errMultipleIPs(interfaceA, addrA, interfaceB, addrB string) error {
        if interfaceA == interfaceB {
                fmt.Errorf("could not choose an IP address to advertise since this system has multiple addresses on interface %s (%s and %s)", interfaceA, addrA, addrB)
        }
        return fmt.Errorf("could not choose an IP address to advertise since this system has multiple addresses on different interfaces (%s on %s and %s on %s)", addrA, interfaceA, addrB, interfaceB)
}
```

**Comment by aboch** [comment]
Thanks @aaronlehmann , rebased. 
I now can now see the new error messages:

```
$ docker swarm init
Error response from daemon: could not choose an IP address to advertise since this system has multiple addresses on different interfaces (10.0.2.15 on eth0 and 192.168.137.250 on eth1) - specify one with --advertise-addr
$ sudo ip addr add dev eth0 5.5.5.5/32
$ docker swarm init
Error response from daemon: could not choose an IP address to advertise since this system has multiple addresses on interface eth0 (10.0.2.15 and 5.5.5.5) - specify one with --advertise-addr
$ sudo ip addr add dev eth0 169.254.5.5
$ sudo ip link set eth1 down
$ docker swarm init
Error response from daemon: could not choose an IP address to advertise since this system has multiple addresses on different interfaces (10.0.2.15 on eth0 and 192.168.137.48 on eth2) - specify one with --advertise-addr
$ sudo ip link set eth2 down
$ docker swarm init
Swarm initialized: current node (41prexn98aicudcfi8xgpq7h2) is now a manager.
<snip>
```

File: daemon/cluster/listen_addr_linux.go (hunk 3 above)
Comment (aboch) [inline daemon/cluster/listen_addr_linux.go]: Let's do it

File: daemon/cluster/listen_addr_linux.go (hunk 2 above)
Comment (aboch) [inline daemon/cluster/listen_addr_linux.go]: I agree that from an algorithm point of view, one may want to always get a result.
But I think in this case is more important to take into account the context in which we are trying to elect an address: advertising our address to a remote node, or binding to an address reachable from a remote node.

In my opinion the documentation or the error message could be updated to reflect this, that we are looking for a remote reachable physical device.

While a machine with no IP on phy devices and only a bridge interface can remotely be reached, it is not really a common setup, it requires some advanced configurations, compared to where, I think, people commonly test swarm: VMs, droplets, laptops, etc...

Also we would end up choosing a docker managed bridge, and I do not want to rely on the subnet provider logic to skip it, given it does not, and cannot take into account that a returned subnet may not even be visible into the host network namespace.

**Comment by aaronlehmann** [comment]
LGTM

**Comment by tonistiigi** [comment]
LGTM

**Comment by thaJeztah** [comment]
ping @aluzzardi @vieux wdyt; 1.13 or 1.12.1 ? (https://github.com/docker/docker/pull/25279#issuecomment-236633825)

**Comment by thaJeztah** [comment]
all green, merging, but

ping @tiborvass wether or not to include in 1.12.1

**Comment by tiborvass** [comment]
Ping @justincormack does this look good to you too?

**Comment by aaronlehmann** [comment]
@aboch: I think this change caused a regression as far as ignoring Docker-specific interfaces like `docker0`.

Using `dind`, with the following interfaces:

```
docker0   Link encap:Ethernet  HWaddr 02:42:71:a6:a4:40
          inet addr:172.18.0.1  Bcast:0.0.0.0  Mask:255.255.0.0
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

eth0      Link encap:Ethernet  HWaddr 02:42:ac:11:00:03
          inet addr:172.17.0.3  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe11:3/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:8 errors:0 dropped:0 overruns:0 frame:0
          TX packets:8 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:648 (648.0 B)  TX bytes:648 (648.0 B)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
```

I get an error from `docker swarm join`:

```
Error response from daemon: could not find the system's IP address - specify one with --advertise-addr
```

This used to correctly ignore `docker0`.

**Comment by justincormack** [comment]
At a guess a veth endpoint is not a "device" as far as netlink is
concerned, so it sees no interfaces.
